Feature/importvalue

filippomc opened the pull request:
We are still in a WIP but there are some important choices which is important to agreed on before going on with this.

What we have here is the PointerUtility and related tests ported from Java with this tool: https://github.com/natural/java2python/ and adapted.

Points worth discussing:
* Use of the porting tool. Porting code is being useful but some implementation/style choices must be agreed:
   * CamelCase from methods: this is the way methods are defined in Java, use of PEP8 standard slows down things a lot. We may leave camel case as a way to say "this was imported from java"
   * Getters and Setters. In Java we're always using getters and setters instead of accessing variables. Unfortunately we have no getters and setters generated from pyecore and that slows down the porting
   * Method overloading: a way to implement overloading is coming from j2py which I like more than functools.singledispatch because it implements multiple dispatching. 
* Python version support: I'm finding type hints really useful when matching the code from Java. Unfortunately they are available only from Python 3.5. 
* OSGI service discovery replacement. I've started with a simple implementation in which we are manually adding services (see model_interpreter.py). Basically when you import a module which defines a new model interpreter, you can add that model interpreter when loading the module.

aranega committed:
Add quick experiment for speed improvement

aranega committed:
Fix missing import in the experiment

filippomc committed:
Started lazy loading refactor

filippomc committed:
lazy loading backend implementation

filippomc committed:
Started implementing a test and stubs for importValue flow

filippomc committed:
Implemented PointerUtility

The implementation was made through a java to python conversion tool.
See https://github.com/natural/java2python/

Unit tests were also ported from java: they're passing with this commit

aranega commented:
Hi @filippometacell

Indeed, as you saw, the Python world is a little different from the Java one

* CamelCase. The usage of camel case is often discouraged in Python for method/functions name. It is prefered to be used for class name only. We could derive from PEP-8 if required.

* Getters and setters are really considered as a bad practice in Python. One of the main use of getter/setter is to provide a stable API. In Python, other mechanism exists (usage of `@property`) to provide the same mechanism. The idea of PyEcore is to provide a pythonic implementation of EMF, so getter/setter are not generated. A thing that could be done is to create a dedicated generator for pygeppetto so the generated code is less pythonic and more javaic. 
* There is dedicated python libraries that allows multiple dispatch and that could fit the job

* Indeed type hint can be useful. From my experience, I used them a lot in the begining, but after a while, they are not useful at all when everything is well named (I'm in the Smalltalk community also, it's also strongly dynamically typed, and everything is fine :) ). But I understand that they can be useful for porting.

* I'm not very familiar with OSGI, but I know there is some equivalent for python or there is mechanisms that can easily be used as the one you provided.

I think at the end, the question is more: Do you want a port in Python, or do you want a Pythonic implementation of Geppetto. The difference is quite a huge one. You can watch the intervention of Raymond Hettinger named "Beyond the PEP-8" that gives an interesting talk about the difference between then two.

I'll read your code :) Thanks a lot @filippometacell  and sorry for my incomplete work :\

aranega commented on pygeppetto/data_model.py:
```diff
@@ -0,0 +1,28 @@
+
```
It's related to what we are discussing, getter and setter are bad practice in Python.

When I started to write the port, I actually rewrote the code for it to be Pythonic and to remove/change some parts.

aranega commented on pygeppetto/model/utils/pointer_utility.py:
```diff
@@ -0,0 +1,347 @@
+from ..exceptions import GeppettoModelException
```
I know it's generated code... but it's really non pythonic and the logic could be improved sometimes :p

filippomc replied:
Agree. I don't like many for loops in particular. It may need time though to make everything pythonic. And we have a lot of things to port there :/

aranega replied:
Ok, if you have a lot of things to port, then the best thing is to use automatic translation as you did and non pythonic code for a first run, as long as tests are here. We could refactor everything later.

I was very slow when I started everything because I was rewriting everything (removing some non-javaic code also from the Java version :p). To my opinion, it's the best thing to do, but it require extra-time and a lot of hands to do it properly.

filippomc replied:
I like the approach to make some refactoring before merging. I'll also keep in mind to do small pythonic refactorings whenever possible also during the porting

filippomc replied:
I think it's much more Pythonic now. Still we have some ugly loops but I wouldn't go in the depth of complex algorithms just for the sake of the style.
I kept the camelCase: I'm not yet convinced about the snake_case at any cost; it's not only saving time during the porting but also keeping the correspondence with Java and Javascript where it comes to API methods, which has a great value in my opinion. What about a style guide deriving from PEP-8 with camel case encouraged only where there is a correspondence with the Java/Javascript side?

aranega commented on tests/test_importvalue.py:
```diff
@@ -0,0 +1,39 @@
+import unittest
```
Other tests have been written using `pytest` which is more powerful and make tests easier to write. Perhaps it would be better to use it or to switch everything to `unittest` so everything is on the same page.

filippomc replied:
It's easier to use unittest when it comes to imported code, since it's more similar to JUnit. pytest import seems broken though, maybe it's missing from the setup? 
I'd go with both since unittest is a standard library so no extra dependencies are involved. Anyway for you the final word!

aranega replied:
The `pytest` dependency is here, but it's an extra one only for tests (in the `setup.py`). Everything can be easily launch with `tox`. I think pytest is compatible with `unittest` so it should not be a problem. Beside, once tests are written, it's easy to rewrite them ;). I perform that for another project and the conversion was not so painful, so it's probably better to use `unittest` for now if it speedup the port and to change later if required.

filippometacell commented:
Hi @aranega, yes I was aware of the bad practices in there, that's why I felt important to share that to you. I always use PEP-8 as a reference and prefer pythonic code, though here we have here are some trade-offs which may favour a less strict interpretation of style guidelines when it comes to geppetto, which is clearly a porting instead.

> * CamelCase. The usage of camel case is often discouraged in Python for method/functions name. It is prefered to be used for class name only. We could derive from PEP-8 if required.
Agree the principle, still is very annoying when it comes to Geppetto ported stuff to have the names changed.

> * Getters and setters are really considered as a bad practice in Python. One of the main use of getter/setter is to provide a stable API. In Python, other mechanism exists (usage of `@property`) to provide the same mechanism. The idea of PyEcore is to provide a pythonic implementation of EMF, so getter/setter are not generated. A thing that could be done is to create a dedicated generator for pygeppetto so the generated code is less pythonic and more javaic.

I'm convinced it's not worth changing PyEcore. This will only require some more work on the porting mainly for the setters.

> * There is dedicated python libraries that allows multiple dispatch and that could fit the job
Yes I wasn't sure about adding external libraries. what about https://pypi.org/project/multimethod/

> * Indeed type hint can be useful. From my experience, I used them a lot in the begining, but after a while, they are not useful at all when everything is well named (I'm in the Smalltalk community also, it's also strongly dynamically typed, and everything is fine :) ). But I understand that they can be useful for porting.
I also made some projects in Smalltalk years ago, love it! The syntax for parameters there greatly makes clearer their meaning and it's really elegant :)
Surely I don't want to start a discussion of typed vs non-typed here:  the point I think is if we want break support below Python 3.5. If we don't, then we have to remove type hints before the merge...

> * I'm not very familiar with OSGI, but I know there is some equivalent for python or there is mechanisms that can easily be used as the one you provided.

Ok, I'll look for some library for that, if you remember something please let me know!

filippomc committed:
Implemented import value. Tests OK

The logic has changed from Java Geppetto. Here we're using the instance
path instead of the type path.

filippomc committed:
Added tests

filippomc committed:
Import value test moved to pytest

filippomc committed:
Some pythonic style enhancements

filippomc committed:
Added test for serializer

filippomc committed:
Removed custom overload method. Replaced with multimethod

filippomc committed:
Started porting importType logic

filippomc committed:
Added Singleton metaclass to manage manager instances

filippomc committed:
Implemented basic project loading from project url

filippomc committed:
Implemented importValue flow. Integration tests OK.

Only missing the serialization for synched variables.

Commit:
custom serializer (waiting for pyecore update)

aranega commented on pygeppetto/model/model_serializer.py:
```diff
@@ -1,13 +1,29 @@
+        self.onlySerialiseDelta = onlySerialiseDelta
+
+    def to_dict_from_obj(self, obj, options, use_uuid, resource):
+        if hasattr(obj, 'synched'):
```
As we are in Python and with PyEcore, you can directly perform a `obj.synched` ;)

aranega commented on pygeppetto/model/model_serializer.py:
```diff
@@ -1,13 +1,29 @@
+                return { 'synched': True }
+            else:
+                d = super().to_dict_from_obj(obj, options, use_uuid, resource)
+                obj.eSet('synched', True)
```
Same here, you can simply use `obj.synched = True`

aranega commented on tests/test_serializer.py:
```diff
@@ -16,12 +16,36 @@ def testSerializer():
-        geppettoModel, True)) == expected_json_synched
+
+    # FIXME Empty strings attr are converted to None and the serializer removes the attr (( from: pyecore -> xmi.py -> _init_modelroot))
+    geppettoModel.eSet('id', '')
```
you can directly use `geppettoModel.id = ''`

aranega commented on tests/test_serializer.py:
```diff
@@ -16,12 +16,36 @@ def testSerializer():
-    assert json.loads(GeppettoSerializer.serialize(
-        geppettoModel, True)) == expected_json_synched
+
+    # FIXME Empty strings attr are converted to None and the serializer removes the attr (( from: pyecore -> xmi.py -> _init_modelroot))
```
There is an option for the XMI serialization that allows you to keep serialized the default attribute even if they have been set to the default value or a None value: `XMIOptions.SERIALIZE_DEFAULT_VALUES`

aranega commented on pygeppetto/model/model_serializer.py:
```diff
@@ -1,13 +1,29 @@
 
+from pyecore.resources.json import DefaultObjectMapper
+
+class SynchableMapper(DefaultObjectMapper):
```
With the modification I will perform on PyEcore side, the `SynchableMapper` will change a little bit (not so much)

filippomc replied:
Hi @aranega is that modification you mentioned done now?

aranega reviewed (commented):
I made some comments in the code, as I adapt PyEcore code regarding your PR and mixing it with my previous implementation, it will change a little bit the MapperCode. The basic modification will be done in the day and I hope the "deserialization" API exposing also.
Also, in a more general way, as in Python we have late binding, you can directly use `my_obj.my_attr` instead of using `my_objeGet(my_attr)` or  `eSet(...)` :)

filippomc committed:
Fixes and tests - importValue temporarly OK (no cloning)

Commit:
Merge pull request #27 from openworm/feature/importvalue

Feature/importvalue

filippomc committed:
Revised access to common library

tarelli commented:
@filippometacell fantastic work! Can we setup Travis to run whatever tests we have before we merge this? @aranega do you want to give a final look at the changes that would go in with this PR? Thanks both!

filippomc committed:
Merge branch 'feature/importvalue' into nwbdev

Commit:
Add TextVariable for metadata

Commit:
Add tests to pipeline

Commit:
Remove init

filippomc committed:
Merge pull request #28 from openworm/test

Test

filippomc committed:
Added abstract api messaging
